### DISCUSSION TITLE
Print Local variable or parameters' Xref in headless.

### DISCUSSION THREAD

**Discussion by anzosasuke** [NationalSecurityAgency/ghidra#3575] [Q&A] [answered]
I want to try and compare if the two function in the same block use the same parameters in the function. I have set two memset which access 2 different memory buffers, but only one is freed using free function in C. I want to show that one parameter is accessed by both the function{RBP + local_220]. I see in the ghidra GUI at the top of the function, the Xref of local variable shows the addresses ![Screenshot from 2021-11-01 15-11-15](https://user-images.githubusercontent.com/33804183/139736072-880cb466-612f-463f-a921-ee4e417e164c.png) can help me compare the parameters of the two function. But I am not able to print out the Xrefs for the local variable. How do I do that? Also if there is better way to do it(which i think there is), please point me to the right direction?

The code snippets I have tried are 

`from ghidra.app.util import XReferenceUtil

for i in functions:
     print('{} \n'.format(i.getParameters()))
     print('{} \n'.format(i.getLocalVariables()))
`
And ` 
for codeUnit in codeUnits:
 print(XReferenceUtil().getXReferences(codeUnit, -1))`

**Comment by dragonmacher** [comment]
Taking a quick peek at our Listing's `VariableXRefFieldFactory`, it seems like you should be able to get references to the function variables:

```
params = function.getParameters();
vars = function.getLocalVariables();
for each param/var
XReferenceUtils.getVariableRefs(var, ...);
```

**Reply by anzosasuke** [reply]
Thanks for quick reply @dragonmacher  but I tried this before and got the error ` print(XReferenceUtil.getVariableRefs(var, xrefs, offcuts )) TypeError: getVariableRefs(): 1st arg can't be coerced to ghidra.program.model.listing.Variable` so i thought I maybe doing something wrong. The code I tried here is ` var = function.getLocalVariables()

> ```
>  for codeUnit in codeUnits:
>     
>      xrefs = XReferenceUtil().getXReferences(codeUnit, -1)
> 
>     offcuts = XReferenceUtil().getOffcutXReferences(codeUnit, -1)
> 
>     print(XReferenceUtil.getVariableRefs(var, xrefs, offcuts))
> ```
> 
> `

**Comment by dragonmacher** [comment, marked as answer]
In your example you are passing in `codeUnit`.   Instead, you must walk the functions of the program and from each function you can get its parameters and local variables.

```
for each function f in program
    vars = f.getLocalVariables()

    List<Reference> xrefs = new ArrayList<>();
    List<Reference> offcuts = new ArrayList<>();

    for each var v in vars
        XReferenceUtils.getVariableRefs(v, xrefs, offcuts)
```

Things to note: 
- the XReferenceUtils method here does not return a result, but instead places the results in the lists passed to the method
- this version of the method on XReferenceUtils does not work on code units, there is a different method for that
- if you wish for references to to parameters, then you also need to do this work by calling `f.getParameters()` in addition to the call I showed above

**Reply by anzosasuke** [reply]
Yup, I get it now. I was giving the each function in a loop before just forgot to mention. I didn't understand the API page for Xreference which is why I was supplying those values from codeunits. But you made it clear. Thanks so much.
